### PR TITLE
Validate and set ps_component from purl

### PR DESCRIPTION
**Pull request by jobselko**
This PR adds validation between `ps_component` and `purl` to check they match each other and also enables the automatic setting of `ps_component` if a request contains only `purl`.

Closes OSIDB-3410

### CODE REVIEW

File: osidb/tests/factories.py [hunk 1]
```diff
@@ -300,9 +300,6 @@ class Meta:
     )
     ps_module = factory.sequence(lambda n: f"ps-module-{n}")
     ps_component = factory.sequence(lambda n: f"ps-component-{n}")
-    purl = factory.sequence(
-        lambda n: f"pkg:rpm/fedora/ps-component-{n}@7.50.3-1.fc25?arch=i386&distro=fedora-25"
-    )
```
Comment (jobselko) [inline osidb/tests/factories.py:305]: Since `purl` can be blank, I deleted it from the factory to avoid updating every test using `Affect`. If anybody thinks that keeping it here and updating all tests is a better option, I can do it.

File: osidb/api_views.py [hunk 2]
```diff
@@ -840,6 +840,25 @@ class AffectView(
...
+
+        if data and data.get("purl") and not data.get("ps_component"):
+            new_data = data.copy()
+            purl = data["purl"]
+            try:
+                ps_component = PackageURL.from_string(purl).name
+            except Exception as exc:
+                raise ValidationError(f"Invalid purl '{purl}': {exc}")
+            new_data["ps_component"] = ps_component
+            kwargs["data"] = new_data
```
Comment (JakubFrejlach) [inline osidb/api_views.py:858]: I don't think this is the right place to perform such field population or performing any validations. This should be responsibility of either model layer or in this case more probably serializer level. We already perform some field prepopulation on serializer level by overriding the create and update methods there so I would use those as a reference. Also this approach still treats `ps_component` as required field on API which should no longer be true as it will be possible to have it blank in the request no.

**Review by JakubFrejlach — CHANGES_REQUESTED** [review]

File: osidb/api_views.py (hunk 2 above)
Comment (osoukup) [reply osidb/api_views.py:858]: I agree that the view is not a good place for it. Even the serializer is not ideal and the business logic should eventually move out of there but at least for now some of it is there. The validation should ideally be in the model validations and setting the component based on the PURL could eg. be in a pre-save signal or some part of the model save procedure.

Comment (jobselko) [reply osidb/api_views.py:858]: @JakubFrejlach I tried to override `create` in `AffectSerializer`/`AffectPostSerializer`, but without success, `update` itself was okay. If you prefer to use this approach here, I need some help to understand what is wrong with current serializers. I will try to update the API to describe that at least one of `purl` and `ps_component` is required.

Comment (JakubFrejlach) [reply osidb/api_views.py:858]: @jobselko I am not sure, what that override didn't work for you but as @osoukup wrote, doing that on the model layer, you can avoid this whole hustle. Just setting the ps_component not required at the serializer should be enough for you to successfully pass through serializer and then do the rest on the model layer.

File: osidb/tests/factories.py (hunk 1 above)
Comment (jobselko) [reply osidb/tests/factories.py:305]: `ps_component` is generated from `purl`, so I would have to add `purl` into every mentioned test. I think that generating it oppositely would not make sense.

**Review by dmonzonis — APPROVED** [review]
LGTM

File: osidb/models/affect.py [hunk 3]
```diff
@@ -172,6 +172,16 @@ class Meta:
     def __str__(self):
         return str(self.uuid)
 
+    def save(self, *args, **kwargs):
+        if self.purl and not self.ps_component:
+            try:
+                ps_component = PackageURL.from_string(self.purl).name
+            except Exception as exc:
+                raise ValidationError(f"Invalid purl '{self.purl}': {exc}")
```
Comment (osoukup) [inline osidb/models/affect.py:180]: There should be some human-readable affect identification so when a user creates/updates multiple affects and then sees an error it is clear which one was invalid.

File: osidb/models/affect.py [hunk 4]
```diff
@@ -483,6 +493,20 @@ def _validate_purl(self, **kwargs):
...
             raise ValidationError(f"Invalid purl '{self.purl}': {exc}")
 
+    def _validate_purl_and_ps_component(self, **kwargs):
+        """
+        Checks that ps_component matches the one included in purl if purl exists.
+        """
+        if (
+            self.purl
+            and self.ps_component
+            and self.ps_component != PackageURL.from_string(self.purl).name
```
Comment (osoukup) [inline osidb/models/affect.py:503]: In the case user provides both but the PURL is invalid the conversion here would result in an exception which would most probably be unclear to the user and without any affect identification. We should catch it and transform into a better one.

File: osidb/models/affect.py [hunk 5]
```diff
@@ -483,6 +493,20 @@ def _validate_purl(self, **kwargs):
...
+        ):
+            raise ValidationError(
+                "ps_component does not match the one included in purl: "
+                f"ps_component: {self.ps_component}, purl: {self.purl}."
```
Comment (osoukup) [inline osidb/models/affect.py:507]: There should be the full affect identification (also PS module) so the user easily sees which one is incorrect.

File: osidb/serializer.py [hunk 6]
```diff
@@ -1035,6 +1035,13 @@ class AffectSerializer(
     trackers = serializers.SerializerMethodField()
     meta_attr = serializers.SerializerMethodField()
     cvss_scores = AffectCVSSSerializer(many=True, read_only=True)
+    # ps_component is not required because it is created in Affect.save()
```
Comment (osoukup) [inline osidb/serializer.py:1038]: ... created out of PURL which is then required. Where do we enforce it?

**Review by osoukup — CHANGES_REQUESTED** [review]
This looks almost OK but I would like the alert messages to be improved. Also it is not clear to me what happens when neither the component (which is no optional) nor PURL is provided. Will uniqueness constrain prevent this to be stored in the database? Will the error be human-readable? I would like to see a test for this case to prevent any potential future regression.

File: osidb/models/affect.py [hunk 7]
```diff
@@ -172,6 +172,26 @@ class Meta:
     def __str__(self):
         return str(self.uuid)
 
+    def save(self, *args, **kwargs):
+        # Check that either purl or ps_component exists
+        if not self.purl and not self.ps_component:
+            raise ValidationError(
+                f"Affect ({self.uuid}) for {self.ps_module} must have either purl or ps_component."
+            )
```
Comment (osoukup) [inline osidb/models/affect.py:180]: This is not the right place for the validation. A validation method should be defined for it

File: osidb/models/affect.py [hunk 8]
```diff
@@ -172,6 +172,26 @@ class Meta:
...
+
+        # Check that the purl field can be parsed into a PackageURL object
+        if self.purl and not self.ps_component:
+            try:
+                ps_component = PackageURL.from_string(self.purl).name
+            except Exception as exc:
+                raise ValidationError(
+                    f"Affect ({self.uuid}) for {self.ps_module} has an invalid purl "
+                    f"'{self.purl}': {exc}. Provide either a valid one or ps_component."
+                )
```
Comment (osoukup) [inline osidb/models/affect.py:190]: The same here. This should be a validation

File: osidb/models/affect.py (hunk 4 above)
Comment (osoukup) [reply osidb/models/affect.py:503]: I was asked to re-review but this does not seem to be addressed. I the PURL string is no parse-able the validation will produce an error message which is unrelated to what is being validated here

**Review by osoukup — CHANGES_REQUESTED** [review]
A bit better but I still do not think that the validations are correct or on the correct place

File: osidb/models/affect.py (hunk 4 above)
Comment (osoukup) [reply osidb/models/affect.py]: I was not referring to the missing tests but to the non-caught PURL parsing exception in the condition

File: osidb/models/affect.py [hunk 9]
```diff
@@ -473,15 +482,31 @@ def _validate_special_consideration_flaw(self, **kwargs):
...
+        elif self.purl and not self.ps_component:
+            try:
+                self.ps_component = PackageURL.from_string(self.purl).name
+            except Exception as exc:
+                raise ValidationError(
+                    f"Affect ({self.uuid}) for {self.ps_module} has an invalid purl "
+                    f"'{self.purl}': {exc}. Provide either a valid one or ps_component."
+                )
+
+        elif self.purl and self.ps_component != PackageURL.from_string(self.purl).name:
```
Comment (osoukup) [inline osidb/models/affect.py:505]: Here is still a non-caught exception possibly happening during the PURL parsing which would result in potentially non-user-readable error message out of our control.

**Review by osoukup — CHANGES_REQUESTED** [review]
Thank you for moving the validation logic to the validations. The non-caught exception is however still there. This seems to be the very last issue I have here

File: osidb/mixins.py [hunk 10]
```diff
@@ -809,6 +805,10 @@ def validate(self, raise_validation_error=True, dry_run=False):
                         **(e.params or {}),
                     )
 
+        # standard validations
+        # exclude meta attributes
+        self.full_clean(exclude=["meta_attr"])
```
Comment (osoukup) [inline osidb/mixins.py:810]: This change should be done in a separate commit as it does not directly relate to PS components or PURLs but with the fact that our custom validations should get priority before the standard ones (as they are more specific and provide better readable and more targeted error messages).

**Review by osoukup — APPROVED** [review]
LGTM
